src/helpers/presentation.rs

```diff
@@ -0,0 +1,323 @@
+    ) -> impl Fn(serde_json::Error) -> PresentationError + 'a {
+        move |err| {
+            PresentationError::JsonDeserialization(
+                to_parse_name.to_string(),
+                err.to_string(),
+                to_parse_value.to_string(),
+            )
+        }
+    }
+}
+
+// Master secret is always incorporated, without being mentioned in the credential schema
+const ADDITIONAL_HIDDEN_MESSAGES_COUNT: usize = 1;
+const NQUAD_REGEX: &str = r"^_:c14n0 <http://schema.org/([^>]+?)>";
+const TYPE_OPTIONS: &str = r#"{ "type": "bbs" }"#;
+
+pub struct Presentation<'a> {
+    vade_evan: &'a mut VadeEvan,
+}
+
+impl<'a> Presentation<'a> {
+    pub fn new(vade_evan: &'a mut VadeEvan) -> Result<Self, PresentationError> {
+        Ok(Self { vade_evan })
+    }
+
+    pub async fn create_proof_request(
+        &mut self,
+        schema_did: &str,
+        revealed_attributes: Option<&str>,
+    ) -> Result<String, PresentationError> {
+        let revealed_attributes_parsed: Option<Vec<String>> = revealed_attributes
+            .map(|ras| {
+                serde_json::from_str(ras).map_err(PresentationError::to_deserialization_error(
+                    "revealed attributes",
+                    ras,
+                ))
+            })
+            .transpose()?;
+        let proof_request_payload = RequestProofPayload {
+            verifier_did: None,
+            schemas: vec![schema_did.to_string()],
+            reveal_attributes: self
+                .get_reveal_attributes_map(schema_did, revealed_attributes_parsed)
+                .await?,
+        };
+
+        let proof_request_json = serde_json::to_string(&proof_request_payload).map_err(
+            PresentationError::to_serialization_error("RequestProofPayload"),
+        )?;
+
+        self.vade_evan
+            .vc_zkp_request_proof(EVAN_METHOD, TYPE_OPTIONS, &proof_request_json)
+            .await
+            .map_err(|err| PresentationError::VadeEvanError(err.to_string()))
+    }
+
+    async fn get_did_document<T>(&mut self, did: &str) -> Result<T, PresentationError>
+    where
+        T: DeserializeOwned,
+    {
+        let did_result_str = self
+            .vade_evan
+            .did_resolve(did)
+            .await
+            .map_err(|err| PresentationError::VadeEvanError(err.to_string()))?;
+
+        if did_result_str == "Not Found" {
+            return Err(PresentationError::SchemaNotFound(did.to_string()));
+        }
+
+        let did_result_value: DidDocumentResult<T> = serde_json::from_str(&did_result_str)
+            .map_err(|err| {
+                PresentationError::SchemaInvalid(
+                    did.to_string(),
+                    PresentationError::to_deserialization_error("DID document", &did_result_str)(
+                        err,
+                    )
+                    .to_string(),
+                )
+            })?;
+
+        Ok(did_result_value.did_document)
+    }
+
+    async fn get_reveal_attributes_map(
```

Comment:
Can we rename it to `valide_revaled_attributes_and_get_indicies` or something? :D

Reply:
Hm, quite a mouthful. :D

Thoughts, that went into the current name are:

- we get something, so `get_`
- the property it is assigned to, is called `reveal_attributes`, so use this (weird) name in the middle as the object in the function name, so `_reveal_attributes_`
- it returns a map, so `_map` (but this part could be dropped, was added to make a contrast to the argument `revealed_attributes` as it has a different format)
- validation should be done on the data, and as long as a function exclusively do validation, we usually don't mention it in function names

Open for better suggestions, but still, `valide_revaled_attributes_and_get_indicies` is a bit hard to wield imho. :D

Reply:
Yeha makes sense in general. Maybe we can just add the indicies to the function name? Maybe get_reveal_attributes_indices_map? :D

Reply:
Updated. ^^